src/MultiRegion/cubed_sphere_boundary_conditions.jl

```diff
@@ -99,18 +100,34 @@ end
     end
 end
 
+@inline function fill_cubed_sphere_field_vertical_halo_event!(grid, field)
+    sides, ordered_bcs = permute_boundary_conditions(field.boundary_conditions)
+    reduced_dims = reduced_dimensions(field)
+
+    idx = findfirst(x -> x isa BottomAndTop, sides)
+    side = sides[idx]
+    bc = select_bc(ordered_bcs[idx])
+
+    sz = (grid.Nx + 2grid.Hx, grid.Ny + 2grid.Hy)
+    of = (-grid.Hx, -grid.Hy)
+    kernel! = fill_halo_kernel!(side, bc, grid, sz, of, field.data, reduced_dims)
```

Comment:
```suggestion
    kernels!, bcs = get_boundary_kernels(boundary_conditions, c, grid, loc, indices)
    kernel! = kernels.bottom_and_top
    bcs = bcs.bottom_and_top
```
I think this should go already without going through the hassle of permuting boundary conditions once more

Reply:
Agreed, but then we wouldn’t be filling the vertical halos at the horizontal halo locations ([commit 6cbb6b0](https://github.com/CliMA/Oceananigans.jl/commit/6cbb6b03f7a2ae7e63442ab891612c8a8d2d4172)), right?

Reply:
ok, I get it. Note that how you implemented it is a bit inefficient because we know already everything and we do not need to permute once again the boundary conditions or build a new filling kernel. To make it a bit more efficient while including the horizontal halos you could extend these functions in `fill_halo_kernels.jl`
```julia
size    = fill_halo_size(data, side, indices, bc, loc, grid)
offset  = fill_halo_offset(size, side, indices)
```
or even this function directly
```julia
@inline function fill_halo_kernels(bcs::FieldBoundaryConditions,
                                   data::OffsetArray,
                                   grid::AbstractGrid,
                                   loc, indices)

    arch = architecture(grid)
    sides, ordered_bcs = permute_boundary_conditions(bcs)
    reduced_dimensions = findall(x -> x isa Nothing, loc)
    reduced_dimensions = tuple(reduced_dimensions...)
    names = Tuple(side_name(side) for side in sides)
    kernels! = []

    for task in 1:length(sides)
        side = sides[task]
        bc   = select_bc(ordered_bcs[task])

        size    = fill_halo_size(data, side, indices, bc, loc, grid)
        offset  = fill_halo_offset(size, side, indices)
        kernel! = fill_halo_kernel!(side, bc, grid, size, offset, data, reduced_dimensions)

        push!(kernels!, kernel!)
    end

    kernels! = tuple(kernels!...)

    return NamedTuple{names}(kernels!), NamedTuple{names}(ordered_bcs)
end
```
However, I am not sure how this would play with `@apply_regionally`, i.e. is it inside or outside? 
I think it might be possible to extend these functions either way, but I leave it up to you.